mlflow/store/model_registry/sqlalchemy_store.py

```diff
@@ -337,62 +345,142 @@ def compute_next_token(current_size):
-                .filter(*conditions)
+                filter_query.options(*self._get_eager_registered_model_query_options())
                 .order_by(*parsed_orderby)
                 .limit(max_results_for_query)
             )
             if page_token:
                 query = query.offset(offset)
-            sql_registered_models = query.all()
+            sql_registered_models = session.execute(query).scalars(SqlRegisteredModel).all()
             next_page_token = compute_next_token(len(sql_registered_models))
             rm_entities = [rm.to_mlflow_entity() for rm in sql_registered_models][:max_results]
             return PagedList(rm_entities, next_page_token)
 
+    @classmethod
+    def _get_search_registered_model_filter_query(cls, parsed_filters, dialect):
+        attribute_filters = []
+        tag_filters = []
+        for f in parsed_filters:
+            type_ = f["type"]
+            key = f["key"]
+            comparator = f["comparator"]
+            value = f["value"]
+            if type_ == "attribute":
+                if key not in ("name",):
+                    raise MlflowException(
+                        f"Invalid attribute name: {key}", error_code=INVALID_PARAMETER_VALUE
+                    )
+                if comparator not in ("=", "!=", "LIKE", "ILIKE"):
+                    raise MlflowException(
+                        f"Invalid comparator for attribute: {comparator}",
+                        error_code=INVALID_PARAMETER_VALUE,
+                    )
+                attr = getattr(SqlRegisteredModel, key)
+                f = SearchUtils.get_sql_filter_ops(attr, comparator, dialect)(value)
+                attribute_filters.append(f)
+            elif type_ == "tag":
+                if comparator not in ("=", "!=", "LIKE", "ILIKE"):
+                    raise MlflowException.invalid_parameter_value(
+                        f"Invalid comparator for tag: {comparator}"
+                    )
+                val_filter = SearchUtils.get_sql_filter_ops(
+                    SqlRegisteredModelTag.value, comparator, dialect
+                )(value)
+                key_filter = SearchUtils.get_sql_filter_ops(
+                    SqlRegisteredModelTag.key, "=", dialect
+                )(key)
+                tag_filters.append(sqlalchemy.and_(key_filter, val_filter))
+            else:
+                raise MlflowException(
+                    f"Invalid token type: {type_}", error_code=INVALID_PARAMETER_VALUE
+                )
+
+        rm_query = select(SqlRegisteredModel).filter(*attribute_filters)
+        if tag_filters:
+            tag_filter_query = (
+                select(SqlRegisteredModelTag.name)
+                .filter(sqlalchemy.or_(*tag_filters))
+                .group_by(SqlRegisteredModelTag.name)
+                .having(sqlalchemy.func.count(sqlalchemy.literal(1)) == len(tag_filters))
+                .subquery()
+            )
```

Comment:
Avoid generating multiple table join on tag table when there are multiple tag filters.
This is the similar approach with databricks backend, I think it is a more efficient approach.

Reply:
Cool! Any benchmark?

Reply:
Does this approach really work?

Let's say we only have a registered model tag `{"key": "key", "value": "value"}` and perform a search with `filter_string="tag.key LIKE 'v%' AND tag.key LIKE '%e'"`. `len(tag_filters)` is 2, but `sqlalchemy.func.count(sqlalchemy.literal(1))` can never be larger than 1 since we only have a single registered model tag.

Reply:
We might need to replace `len(tag_filters)` with the number of unique tag keys in `len(tag_filters)`.

Reply:
yes. I updated.